src/mgr/QueryParams.cc

```diff
@@ -66,60 +69,59 @@ Mgr::QueryParams::find(const String& name) const
 }
 
 bool
-Mgr::QueryParams::ParseParam(const String& paramStr, Param& param)
+Mgr::QueryParams::ParseParam(SBuf &paramStr, Param &param)
 {
-    bool parsed = false;
-    regmatch_t pmatch[3];
-    regex_t intExpr;
-    regcomp(&intExpr, "^([a-z][a-z0-9_]*)=([0-9]+((,[0-9]+))*)$", REG_EXTENDED | REG_ICASE);
-    regex_t stringExpr;
-    regcomp(&stringExpr, "^([a-z][a-z0-9_]*)=([^&= ]+)$", REG_EXTENDED | REG_ICASE);
-    if (regexec(&intExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        std::vector<int> array;
-        int n = pmatch[2].rm_so;
-        for (int i = n; i < pmatch[2].rm_eo; ++i) {
-            if (paramStr[i] == ',') {
-                array.push_back(atoi(paramStr.substr(n, i).termedBuf()));
-                n = i + 1;
-            }
-        }
-        if (n < pmatch[2].rm_eo)
-            array.push_back(atoi(paramStr.substr(n, pmatch[2].rm_eo).termedBuf()));
-        param.second = new IntParam(array);
-        parsed = true;
-    } else if (regexec(&stringExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        param.second = new StringParam(paramStr.substr(pmatch[2].rm_so, pmatch[2].rm_eo));
-        parsed = true;
+    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
+
+    Parser::Tokenizer tok(paramStr);
+
+    SBuf name;
+    if (!tok.prefix(name, nameChars))
+        return false;
+
+    if (!tok.skip('='))
+        return false; // TODO: support name-only parameters
+
+    static const CharacterSet badValueChars = CharacterSet("param-value-bad", "&= ");
+    SBuf valueStr = tok.remaining();
+    if (valueStr.findFirstOf(badValueChars) != SBuf::npos)
+        return false;
+
+    param.first = SBufToString(name);
+
+    std::vector<int> array;
+    int64_t intVal = 0;
+    while (tok.int64(intVal, 10, false)) {
+        static const CharacterSet comma("comma",",");
+        (void)tok.skipAll(comma);
+        Must(intVal >= std::numeric_limits<int>::min());
+        Must(intVal <= std::numeric_limits<int>::max());
```

Comment:
Please consider whether this is correct. Without doing additional tests, this causes an exception in some testing, with the following request: https://paste.pr0.tips/Xr?c

Reply:
> this causes an exception in some testing

This code should throw when the raw value cannot be represented as `int`. While I do not recommend using Must() for input validation anymore, I do not see the problem you are alluding to.

Reply:
While the patch I am using it outdated (it is from the first, original patch), I believe the issue is there will be a memory leak if this exception happens. At least in my testing which includes the original patch, a memory leak happens.

I cannot test with the current patch because it does not cleanly patch, and I do not have the time to manually do it.

Reply:
> there will be a memory leak if this exception happens

Aha, _that_ is a bug! In summary: The leak (if any) is a bug. The exception is not a bug. Using Must() for input validation is currently still acceptable, but throwing an explicit exception or providing/using an API dedicated to input validation is where we should be heading long-term.

It is trivial to trigger this exception, so Amos would be able to test whether the current PR code still leaks. Thank you for reminding us about that danger!

Reply:
Ah, sorry; I was still looking at this bug through the lens of "an exception may occur which will may trigger a memory leak", rather than anything specific to the exception itself. I am glad it is clear now.

Reply:
AFAIK none of the data types used for this new parser lead to memory leaks. Our types here have all been designed for exception safe memory handling, others are used as locals to avoid leaks the old allocations had.

Reply:
I will reserve my patches for when this is pushed to master (alternatively, simply send the request to Squid (assuming cachemanager is open) which I posted and check for leaks).

Reply:
The new ParseUrl() sequence _looks_ exception-safe to me, but I agree that this impression should be tested, especially since (minimal) tests are easy to perform in this case.

Reply:
Basic valgrind testing shows no parser leak for the testCacheManager binary which throws on the bad-input tests.

Reply:
FTR, the memory leak I am seeing with the original 3-patches is apparently happening here: https://paste.pr0.tips/3K
with the logs:
http://paste.pr0.tips/qg

I have not tested with the newest code yet.

Reply:
The memory leak goes away with the current full patch. Thanks.